articles/specs/launch.0.1.0.xsd

```diff
@@ -0,0 +1,582 @@
+      </xs:attribute>
+      <xs:attribute name="from" type="xs:string" use="optional">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            Path to the parameters file to be loaded.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+    </xs:complexType>
+  </xs:element>
+
+  <xs:element name="remap">
+    <xs:annotation>
+      <xs:documentation xml:lang="en">
+        Remaps ROS names for a node.
+      </xs:documentation>
+    </xs:annotation>
+
+    <xs:complexType>
+      <xs:attribute name="from" type="xs:string" use="required">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            Name matching expression to look for replacement candidates.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+      <xs:attribute name="to" type="xs:string" use="required">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            Name replacement expression to replace candidates found.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+    </xs:complexType>
+  </xs:element>
+
+  <xs:element name="node">
+    <xs:annotation>
+      <xs:documentation xml:lang="en">
+        Executes a ROS node executable in a local OS process.
+      </xs:documentation>
+    </xs:annotation>
+
+    <xs:complexType>
+      <xs:choice minOccurs="0" maxOccurs="unbounded">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            A collection of ROS parameters, ROS remappings and/or
+            environment variable settings for the launched ROS node.
+          </xs:documentation>
+        </xs:annotation>
+
+        <xs:element ref="env"/>
+        <xs:element ref="param"/>
+        <xs:element ref="params"/>
+        <xs:element ref="remap"/>
+      </xs:choice>
+    </xs:complexType>
+    <xs:attribute name="package" type="xs:string" use="required">
+      <xs:annotation>
+        <xs:documentation xml:lang="en">
+          Name of the package where the node is to be found.
+        </xs:documentation>
+      </xs:annotation>
+    </xs:attribute>
+    <xs:attribute name="executable" type="xs:string" use="required">
+      <xs:annotation>
+        <xs:documentation xml:lang="en">
+          Name of the node executable.
+        </xs:documentation>
+      </xs:annotation>
+    </xs:attribute>
+    <xs:attribute name="name" type="xs:string" use="optional">
+      <xs:annotation>
+        <xs:documentation xml:lang="en">
+          A name for the launched ROS node.
+        </xs:documentation>
+      </xs:annotation>
+    </xs:attribute>
+      <xs:attribute name="args" type="xs:string" use="optional">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">
+            Additional 'command-line' arguments for the ROS node.
+          </xs:documentation>
+        </xs:annotation>
+      </xs:attribute>
+      <xs:attribute name="ns" type="xs:string" use="optional">
```

Comment:
Would it be worthwhile to prevent the use of trailing slashes and perhaps enforce naming constraints? E.g:

``` xml
<xs:simpleType>
  <xs:restriction base="xs:string">
    <xs:pattern value="((\/)?([a-zA-Z_])+([a-zA-Z0-9_])*)+"/>
  </xs:restriction>
</xs:simpleType>
```

https://design.ros2.org/articles/topic_and_service_names.html#ros-2-topic-and-service-name-constraints

Although, I suppose this expression alone doesn't account/guard for the substitution syntax.

Reply:
> Although, I suppose this expression alone doesn't account/guard for the substitution syntax.

Right, that's why I didn't even try to get too specific with the pattern. Same goes everywhere else. Do you think we should?

Reply:
Perhaps we could could use an XML Template to pre-process advanced launch file markup, e.g. expand the substitutions, so that the final expanded XML is a smaller subset of launch file primitives. We could use two schemas, the first would be loose, ensuring the integrity of the input structure as anticipated by the template transform. The second schema (overloaded from the first schema) would validate the output from the transform to additionally ensure the integrity of the input values.

Perhaps this isn't be the best way about this, but I'd like to see the launch syntax remain implementation agnostic, so that it's easier to incorporate into other system orchestration tools. Perhap something like [Xproc](http://xproc.org) could be used to formulate the pipeline.

Reply:
Hmm, substitutions are a core part of the launch system. As in, substitutions have entity beyond the markup realm, be it XML, YAML, TOML, etc., and thus there isn't such a thing as an expanded XML. I guess you could do so though, not sure if we have (or plan to have) substitutions whose value is not fixed at configuration time. @wjwwood I fall back to you with the question.

Reply:
I'd start without markup level (or preprocessing level) substitution, and instead endeavor to have the markup language capture the substitutions, or failing that have the markup ignore the substitutions, leaving them for the launch code to interpret.